File: src-tauri/src/app/modules/repositories/modrinth/modrinth_repository.rs
use super::super::interfaces::IRepository;
use super::super::models::ModResult;
use super::models::SearchResponse;
use crate::app::{
    modules::{
        manifest::models::Side,
        repositories::{
            models::VersionResult,
            modrinth::models::{FindResponse, VersionItem},
        },
    },
    PAGINATION_SIZE,
};
use async_trait::async_trait;
use reqwest::Client;

pub struct ModrinthRepository {
    client: Client,
}

impl ModrinthRepository {
    pub fn new() -> Self {
        Self {
            client: Client::new(),
        }
    }
}

#[async_trait]
impl IRepository for ModrinthRepository {
    async fn search(&self, query: &str, page: usize) -> Vec<ModResult> {
        let url =
            format!("https://api.modrinth.com/v2/search?query={query}") + &with_pagination(page);

        // Send request
        let res = match self.client.get(&url).send().await {
            Ok(r) => r,
            Err(_) => return vec![],
        };

        // Parse JSON
        let parsed: SearchResponse = match res.json().await {
            Ok(p) => p,
            Err(_) => return vec![],
        };

        parsed
            .hits
            .into_iter()
            .map(|hit| ModResult {
                id: hit.project_id.clone(),
                slug: hit.slug.clone(),
                name: hit.title,
                description: hit.description,
                source: "Modrinth".into(),
                side: get_side(&hit.client_side, &hit.server_side),
                url: format!("https://modrinth.com/mod/{}", hit.slug.clone()),
            })
            .collect()
    }

    async fn find(&self, slug: &str) -> Option<ModResult> {
        let url = format!("https://api.modrinth.com/v2/project/{slug}");

        // Send request
        let res = match self.client.get(&url).send().await {
            Ok(r) => r,
            Err(_) => return None,
        };

        // Parse JSON
        let response: FindResponse = match res.json().await {
            Ok(p) => p,
            Err(_) => return None,
        };

        return Some(ModResult {
            id: response.id.clone(),
            slug: response.slug.clone(),
            name: response.title,
            description: response.description,
            source: "Modrinth".into(),
            side: get_side(&response.client_side, &response.server_side),
            url: format!("https://modrinth.com/mod/{}", response.slug.clone()),
        });
    }

    async fn get_versions(
        &self,
        project_id: &str,
        game_versions: &[String],
        loaders: &[String],
    ) -> Vec<VersionResult> {
        let url = format!(
            "https://api.modrinth.com/v2/project/{}/version?game_versions={}&loaders={}",
            project_id,
            serde_json::to_string(game_versions).unwrap(),
            serde_json::to_string(loaders).unwrap(),
        );

        let res = match self.client.get(&url).send().await {
            Ok(r) => r,
            Err(_) => return vec![],
        };

        let parsed: Vec<VersionItem> = match res.json().await {
            Ok(p) => p,
            Err(_) => return vec![],
        };

        parsed
            .into_iter()
            .map(|v: VersionItem| VersionResult {
                mod_id: v.project_id,
                version: v.version_number,
                minecraft_versions: v.game_versions,
                url: v
                    .files
                    .iter()
                    .find(|f| f.primary)
                    .map(|f| f.url.clone())
                    .unwrap_or_default(),
                hash: v
                    .files
                    .iter()
                    .find(|f| f.primary)
                    .and_then(|f| f.hashes.get("sha512").cloned())
                    .or_else(|| {
                        v.files
                            .iter()
                            .find(|f| f.primary)
                            .and_then(|f| f.hashes.get("sha1").cloned())
                    })
                    .unwrap_or("".to_string()),
            })
            .collect()
    }
}

fn get_side(client: &str, server: &str) -> Side {
    let client_supported = matches!(client, "required" | "optional");
    let server_supported = matches!(server, "required" | "optional");

    match (client_supported, server_supported) {
        (true, true) => Side::Both,
        (true, false) => Side::Client,
        (false, true) => Side::Server,
        _ => Side::Unknown,
    }
}

fn with_pagination(page: usize) -> String {
    let offset: usize = page * PAGINATION_SIZE;
    format!("&offset={offset}&limit={PAGINATION_SIZE}")
}

File: packages/core/src/repositories/modrinth/modrinth-repository.ts
// ModrinthRepository ported from src-tauri/src/app/modules/repositories/modrinth/modrinth_repository.rs

import type { IRepository } from "../repository.interface.js";
import type { ModResult, VersionResult } from "../../models/repository.js";
import type { Side } from "../../models/manifest.js";
import type {
  SearchResponse,
  FindResponse,
  VersionItem,
} from "./modrinth-models.js";

const PAGINATION_SIZE = 20;

export class ModrinthRepository implements IRepository {
  private headers: Record<string, string> = {};

  constructor(modrinthToken?: string) {
    if (modrinthToken) {
      this.headers["Authorization"] = `Bearer ${modrinthToken}`;
    }
  }

  async search(query: string, page: number): Promise<ModResult[]> {
    const offset = page * PAGINATION_SIZE;
    const url = `https://api.modrinth.com/v2/search?query=${encodeURIComponent(query)}&offset=${offset}&limit=${PAGINATION_SIZE}`;

    try {
      const res = await fetch(url, { headers: this.headers });
      if (!res.ok) return [];
      const parsed: SearchResponse = await res.json();

      return parsed.hits.map((hit) => ({
        id: hit.project_id,
        slug: hit.slug,
        name: hit.title,
        description: hit.description,
        source: "Modrinth",
        side: getSide(hit.client_side, hit.server_side),
        url: `https://modrinth.com/mod/${hit.slug}`,
      }));
    } catch {
      return [];
    }
  }

  async find(slug: string): Promise<ModResult | undefined> {
    const url = `https://api.modrinth.com/v2/project/${encodeURIComponent(slug)}`;

    try {
      const res = await fetch(url, { headers: this.headers });
      if (!res.ok) return undefined;
      const response: FindResponse = await res.json();

      return {
        id: response.id,
        slug: response.slug,
        name: response.title,
        description: response.description,
        source: "Modrinth",
        side: getSide(response.client_side, response.server_side),
        url: `https://modrinth.com/mod/${response.slug}`,
      };
    } catch {
      return undefined;
    }
  }

  async getVersions(
    projectId: string,
    gameVersions: string[],
    loaders: string[],
  ): Promise<VersionResult[]> {
    const url =
      `https://api.modrinth.com/v2/project/${encodeURIComponent(projectId)}/version` +
      `?game_versions=${encodeURIComponent(JSON.stringify(gameVersions))}` +
      `&loaders=${encodeURIComponent(JSON.stringify(loaders))}`;

    try {
      const res = await fetch(url, { headers: this.headers });
      if (!res.ok) return [];
      const parsed: VersionItem[] = await res.json();

      return parsed.map((v) => {
        const primaryFile = v.files.find((f) => f.primary);
        return {
          modId: v.project_id,
          version: v.version_number,
          minecraftVersions: v.game_versions,
          url: primaryFile?.url ?? "",
          hash:
            primaryFile?.hashes?.sha512 ??
            primaryFile?.hashes?.sha1 ??
            "",
        };
      });
    } catch {
      return [];
    }
  }
}

function getSide(client: string, server: string): Side {
  const clientSupported = client === "required" || client === "optional";
  const serverSupported = server === "required" || server === "optional";

  if (clientSupported && serverSupported) return "both";
  if (clientSupported) return "client";
  if (serverSupported) return "server";
  return "unknown";
}
